asQ/allatonce/form.py

```diff
@@ -59,8 +58,9 @@ def __init__(self,
         for bc in self.bcs:
             bc.apply(aaofunc.function)
 
-        # function to assemble the nonlinear residual into
-        self.F = aaofunc.copy(copy_values=False).zero()
+        # cofunction to assemble the nonlinear residual into
+        self.F = AllAtOnceCofunction(self.ensemble, self.time_partition,
+                                     aaofunc.field_function_space.dual())
```

Comment:
What's the reason for this change?

Reply:
`firedrake.assemble` now only accepts a `Cofunction` as the `tensor` argument. Before we were passing a `Function` which was the internal buffer for the residual `aaoform.F`, i.e. an `AllAtOnceFunction`. Now I have changed the aaform's residual buffer to an `AllAtOnceCofunction`, so that we can pass a `firedrake.Cofunction` to `firedrake.assemble` instead.

So now we should be consistent with Firedrake in that residuals are always `(AllAtOnce)Cofunctions`